Reimplement NodeHoisting Transformation

phate opened the pull request:
This PR implements a new algorithm for the node hoisting transformation. The algorithm avoids the runtime issues of the old algorithm by not copying nodes to every legal path in the region tree, but only to its final target region. It accomplishes this in a two step process:

1. Marks every node with its target region where it can be hoisted.
2. Performs the hoisting.

phate commented on jlm/llvm/opt/push.cpp:
```diff
@@ -47,412 +49,388 @@ class NodeHoisting::Statistics final : public util::Statistics
-      auto tmp = rvsdg::TryGetOwnerNode<rvsdg::Node>(user);
-      if (tmp && !hasPredecessors(*tmp) && is_theta_invariant(tmp, invariants))
-        wl.push_back(tmp);
+      // FIXME: We currently need to leave control constants where they are even though there is
```
@caleridas Is there still any reason (was there ever?) why we model the ControlType with as a state kind? I would rather like to convert it to a value kind to avoid these problems.

haved reviewed (dismissed):
@phate very nice, just tiny comments

haved commented:
@phate does this code end up moving thetas out of gammas if they do not have side effects? My impression is that it does, but that the theta is then supposed to be moved back in later by the pull in pass. Is this understanding correct?

phate commented on jlm/llvm/opt/push.cpp:
```diff
@@ -47,412 +49,388 @@ class NodeHoisting::Statistics final : public util::Statistics
   {
-    JLM_ASSERT(queue_.size() == set_.size());
-    return queue_.empty();
+    JLM_ASSERT(RegionDepth_.find(&region) != RegionDepth_.end());
```
Removed

phate commented:
@haved I need a review again. I fixed the handling of nullary nodes. ControlConstantOperation is not the only nullary node we need to take care of. We also need to handle UndefValueOperation etc. I changed the code to generally handle it.
Please have a look again.

Regarding theta nodes: No, theta and gamma nodes are not moved at all (for now). There is a FIXME for that on line 257 in push.cpp.
